quantize: correct results when rounding

mjibson opened the pull request:
Not sure how we passed the GDA tests before, but the first test case
added in TestQuantize was failing.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/apd/26)
<!-- Reviewable:end -->

RaduBerinde commented:
Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions.

---

*[context.go, line 934 at r1](https://reviewable.io:443/reviews/cockroachdb/apd/26#-KdlYwMKMII3mINg1bhx:-KdlYwMKMII3mINg1bhy:b1xm4k) ([raw file](https://github.com/cockroachdb/apd/blob/190244f232f176a63c917a5986c65ff9ef4d9020/context.go#L934)):*
> ```Go
> }
> 
> func (c *Context) quantize(d, v, e *Decimal) Condition {
> ```

add a comment explaining what this does. Also why do we need `e` to be a `*Decimal`? Seems like we could just pass the exponent as an int32 or whatever

---

*[context.go, line 955 at r1](https://reviewable.io:443/reviews/cockroachdb/apd/26#-KdlZtLlV4M7RHJzA263:-KdlZtLmdZ1UXb2hO6Aj:bpcdkjp) ([raw file](https://github.com/cockroachdb/apd/blob/190244f232f176a63c917a5986c65ff9ef4d9020/context.go#L955)):*
> ```Go
> 			nc := c.WithPrecision(uint32(p))
> 			neg := d.Sign() < 0
> 			d.Exponent = -diff
> ```

`d.Exponent` here is effectively just a temporary variable which just makes things confusing.

---

*[context.go, line 961 at r1](https://reviewable.io:443/reviews/cockroachdb/apd/26#-Kdl_Kmz13BheEJ7cS7y:-Kdl_Kn-5e-RvNtMDh6g:bc92ts8) ([raw file](https://github.com/cockroachdb/apd/blob/190244f232f176a63c917a5986c65ff9ef4d9020/context.go#L961)):*
> ```Go
> 			for d.Exponent > 0 {
> 				d.Exponent--
> 				d.Coeff.Mul(&d.Coeff, bigTen)
> ```

I'm a little confused why we want to do this. Isn't 100e-4 equivalent to 1e-2? And isn't the latter preferred for efficiency?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/apd/26)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions.

---

*[context.go, line 934 at r1](https://reviewable.io:443/reviews/cockroachdb/apd/26#-KdlYwMKMII3mINg1bhx:-Kdm3tZVj-6oYM241c0z:b-896fix) ([raw file](https://github.com/cockroachdb/apd/blob/190244f232f176a63c917a5986c65ff9ef4d9020/context.go#L934)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

add a comment explaining what this does. Also why do we need `e` to be a `*Decimal`? Seems like we could just pass the exponent as an int32 or whatever
</blockquote></details>

Done.

---

*[context.go, line 955 at r1](https://reviewable.io:443/reviews/cockroachdb/apd/26#-KdlZtLlV4M7RHJzA263:-Kdm3vdOM8mQ0Dl5tvyP:b8a80h4) ([raw file](https://github.com/cockroachdb/apd/blob/190244f232f176a63c917a5986c65ff9ef4d9020/context.go#L955)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

`d.Exponent` here is effectively just a temporary variable which just makes things confusing.
</blockquote></details>

The idea here is that the resulting d.Exponent after rounding will be 0. This is because we have a number of, say, 5 digits, but p (our precision) above is set a 3. So here d.Exponent is set to `-2`. We have a number like `NNN.xx`, where the `.xx` part will be rounded away. However during rounding of 0.9 to 1.0, d.Exponent will be set to 1 instead of 0, so we have to reduce it and increase the coefficient below. It's not quite a temporary variable.

I've added this comment to try to explain my reasoning. (However it's possible my reasoning or implementation is incorrect.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/apd/26)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[context.go, line 955 at r1](https://reviewable.io:443/reviews/cockroachdb/apd/26#-KdlZtLlV4M7RHJzA263:-KdmEYWeZ7ecAnNsoLBR:bjfuetr) ([raw file](https://github.com/cockroachdb/apd/blob/190244f232f176a63c917a5986c65ff9ef4d9020/context.go#L955)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

The idea here is that the resulting d.Exponent after rounding will be 0. This is because we have a number of, say, 5 digits, but p (our precision) above is set a 3. So here d.Exponent is set to `-2`. We have a number like `NNN.xx`, where the `.xx` part will be rounded away. However during rounding of 0.9 to 1.0, d.Exponent will be set to 1 instead of 0, so we have to reduce it and increase the coefficient below. It's not quite a temporary variable.

I've added this comment to try to explain my reasoning. (However it's possible my reasoning or implementation is incorrect.)
</blockquote></details>

Ah. I missed the fact that `Round` uses and changes `d.Exponent`.

What I don't understand then is why we are setting the exponent to `-diff` and then trying to make it 0 instead of letting it be what it is, rounding to the calculated precision, and then adjusting to make sure it's `exp`? (I realize these are equivalent, but for clarity)

---

*[context.go, line 961 at r1](https://reviewable.io:443/reviews/cockroachdb/apd/26#-Kdl_Kmz13BheEJ7cS7y:-KdmGCyzHhrihthTysWN:b-shavap) ([raw file](https://github.com/cockroachdb/apd/blob/190244f232f176a63c917a5986c65ff9ef4d9020/context.go#L961)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I'm a little confused why we want to do this. Isn't 100e-4 equivalent to 1e-2? And isn't the latter preferred for efficiency?
</blockquote></details>

This loop could have a lot of iterations no? e.g. for 0.99999999 -> 1.0 Why not mul with `tableExp10` like above? Better yet - move this whole branch to the top, and let the other branch handle this, something like:
```
if d.Exponent > exp {
   // round (which brings the exponent to at most `exp`).
}
if d.Exponent < exp {
  ..
}
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/apd/26)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 0 of 2 files reviewed at latest revision, 4 unresolved discussions.

---

*[context.go, line 925 at r2](https://reviewable.io:443/reviews/cockroachdb/apd/26#-KdmITt1mKUEwIkDEOwu:-KdmITt1mKUEwIkDEOwv:b-nqcavb) ([raw file](https://github.com/cockroachdb/apd/blob/1e19da4d32b9f78ab87f60e290e7273788161251/context.go#L925)):*
> ```Go
> 
> // Quantize sets d to the value of v with x's Exponent.
> func (c *Context) Quantize(d, v, x *Decimal) (Condition, error) {
> ```

can't we change this to take an exponent too?

---

*[context.go, line 934 at r2](https://reviewable.io:443/reviews/cockroachdb/apd/26#-KdmIWSF-f1pF1a7_jNP:-KdmIWSF-f1pF1a7_jNQ:bj2xgw5) ([raw file](https://github.com/cockroachdb/apd/blob/1e19da4d32b9f78ab87f60e290e7273788161251/context.go#L934)):*
> ```Go
> }
> 
> // quantize sets d to the value of v with exponent exp.
> ```

this description is still not 100% clear.. maybe "adjusts and rounds v as necessary so it is represented with exponent exp and stores the result in `d` (same above). It's fine if we just comment the function above (I didn't notice it earlier in reviewboard)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/apd/26)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Reviewed 1 of 2 files at r1.
Review status: 1 of 2 files reviewed at latest revision, 6 unresolved discussions.

---

*[context.go, line 925 at r2](https://reviewable.io:443/reviews/cockroachdb/apd/26#-KdmITt1mKUEwIkDEOwu:-KdnnL5Ucoe4efbA1EYX:bkw2pie) ([raw file](https://github.com/cockroachdb/apd/blob/1e19da4d32b9f78ab87f60e290e7273788161251/context.go#L925)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

can't we change this to take an exponent too?
</blockquote></details>

+1 if this restrictive interface ever forces us to allocate unnecessarily.

---

*[context.go, line 973 at r2](https://reviewable.io:443/reviews/cockroachdb/apd/26#-KdnpRtvRaCaPCRx1SuW:-KdnpRtvRaCaPCRx1SuX:b-vqxooy) ([raw file](https://github.com/cockroachdb/apd/blob/1e19da4d32b9f78ab87f60e290e7273788161251/context.go#L973)):*
> ```Go
> 			// unexpectedly removes a negative sign when converting from -9 to -10. This
> 			// check is needed until it is fixed.
> 			if neg && d.Sign() > 0 {
> ```

Can we log something here or panic? If there's a bug, it would be nice to catch it ASAP.

---

*[decimal_test.go, line 428 at r2](https://reviewable.io:443/reviews/cockroachdb/apd/26#-KdnoXrrXTIgZhXCF3Xs:-KdnoXrsuX1wuIOC3MTX:b-fmz4my) ([raw file](https://github.com/cockroachdb/apd/blob/1e19da4d32b9f78ab87f60e290e7273788161251/decimal_test.go#L428)):*
> ```Go
> 	tests := []struct {
> 		s      string
> 		e      int32
> ```

no tests for `e > 0` or `e == 0`? Also, no tests for `|e| >= len(s)`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/apd/26)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Review status: 0 of 4 files reviewed at latest revision, 6 unresolved discussions.

---

*[context.go, line 955 at r1](https://reviewable.io:443/reviews/cockroachdb/apd/26#-KdlZtLlV4M7RHJzA263:-KdpHeCrig6klzKOqO2S:b-ipjfl4) ([raw file](https://github.com/cockroachdb/apd/blob/190244f232f176a63c917a5986c65ff9ef4d9020/context.go#L955)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Ah. I missed the fact that `Round` uses and changes `d.Exponent`.

What I don't understand then is why we are setting the exponent to `-diff` and then trying to make it 0 instead of letting it be what it is, rounding to the calculated precision, and then adjusting to make sure it's `exp`? (I realize these are equivalent, but for clarity)
</blockquote></details>

I've added another comment about why we do it this way instead of the way you describe. It's because the way you describe raises Underflow flags and basing around 0 doesn't, and the Quantize spec says it doesn't raise Underflow. Specifically, it fails the quax545 test (https://github.com/cockroachdb/apd/blob/master/testdata/quantize.decTest#L613) with incorrect flags `expected flags "inexact"; got flags "underflow, inexact, subnormal, clamped"`.

---

*[context.go, line 961 at r1](https://reviewable.io:443/reviews/cockroachdb/apd/26#-Kdl_Kmz13BheEJ7cS7y:-KdpIDaHtrZxfB9Uv-gZ:b-ucoi1x) ([raw file](https://github.com/cockroachdb/apd/blob/190244f232f176a63c917a5986c65ff9ef4d9020/context.go#L961)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

This loop could have a lot of iterations no? e.g. for 0.99999999 -> 1.0 Why not mul with `tableExp10` like above? Better yet - move this whole branch to the top, and let the other branch handle this, something like:
```
if d.Exponent > exp {
   // round (which brings the exponent to at most `exp`).
}
if d.Exponent < exp {
  ..
}
```
</blockquote></details>

The api for this function is that the resulting `d.Exponent == exp`, so we must adjust d.Coeff to be whatever it needs to in order to get that. Adding zeros to the coeff is the only way to do that.

The 0.99999 -> 1.0 case will still only loop one time in the for loop. Although it was a for loop, it should only execute at most 1 times. I've changed it to an if statement, and the tests still pass.

---

*[context.go, line 925 at r2](https://reviewable.io:443/reviews/cockroachdb/apd/26#-KdmITt1mKUEwIkDEOwu:-KdpFCeHsUZ8-PSk6vV7:b-896fix) ([raw file](https://github.com/cockroachdb/apd/blob/1e19da4d32b9f78ab87f60e290e7273788161251/context.go#L925)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

+1 if this restrictive interface ever forces us to allocate unnecessarily.
</blockquote></details>

Done.

---

*[context.go, line 934 at r2](https://reviewable.io:443/reviews/cockroachdb/apd/26#-KdmIWSF-f1pF1a7_jNP:-KdpFgy-jQd1uPjYBhze:b-5fj118) ([raw file](https://github.com/cockroachdb/apd/blob/1e19da4d32b9f78ab87f60e290e7273788161251/context.go#L934)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

this description is still not 100% clear.. maybe "adjusts and rounds v as necessary so it is represented with exponent exp and stores the result in `d` (same above). It's fine if we just comment the function above (I didn't notice it earlier in reviewboard)
</blockquote></details>

Done. This function is indeed difficult to describe.

---

*[context.go, line 973 at r2](https://reviewable.io:443/reviews/cockroachdb/apd/26#-KdnpRtvRaCaPCRx1SuW:-KdpMp3EnYbh-9AYyIgF:bxxdsgb) ([raw file](https://github.com/cockroachdb/apd/blob/1e19da4d32b9f78ab87f60e290e7273788161251/context.go#L973)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Can we log something here or panic? If there's a bug, it would be nice to catch it ASAP.
</blockquote></details>

I have another PR out to just fix this bug.

---

*[decimal_test.go, line 428 at r2](https://reviewable.io:443/reviews/cockroachdb/apd/26#-KdnoXrrXTIgZhXCF3Xs:-KdpJKGqnk_TW9rCanI3:bb0xb8l) ([raw file](https://github.com/cockroachdb/apd/blob/1e19da4d32b9f78ab87f60e290e7273788161251/decimal_test.go#L428)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

no tests for `e > 0` or `e == 0`? Also, no tests for `|e| >= len(s)`?
</blockquote></details>

The tests here are just a supplement to the GDA tests, and not meant to be exhaustive. I found a single test case that failed and wasn't also covered by GDA (the 1.00 test). 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/apd/26)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 4 files reviewed at latest revision, 6 unresolved discussions.

---

*[context.go, line 961 at r1](https://reviewable.io:443/reviews/cockroachdb/apd/26#-Kdl_Kmz13BheEJ7cS7y:-KdphCBiUr7RL7Kcyc3G:b-wcpdx4) ([raw file](https://github.com/cockroachdb/apd/blob/190244f232f176a63c917a5986c65ff9ef4d9020/context.go#L961)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

The api for this function is that the resulting `d.Exponent == exp`, so we must adjust d.Coeff to be whatever it needs to in order to get that. Adding zeros to the coeff is the only way to do that.

The 0.99999 -> 1.0 case will still only loop one time in the for loop. Although it was a for loop, it should only execute at most 1 times. I've changed it to an if statement, and the tests still pass.
</blockquote></details>

Weird, I would have expected that rounding 0.99999 would have given us 1 with exponent 0. But I see that this is working. Since we now have an `if`, can we add a check with panic that the exponent is 1? (just so it's clear)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/apd/26#-:-KdpjXteSH10b_UlD2Gr:bnfp4nl)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 4 files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/apd/26#-:-KdsIz0XmsTdv_Bshg6p:bnfp4nl)*
<!-- Sent from Reviewable.io -->
